docs/plugins.md

```diff
@@ -0,0 +1,72 @@
+# Plugins
+
+![Plugins](assets/plugins.png)
+
+Plugins allow you to extend Argo Workflows to add new capabilities.
+
+* You don't need to learn Golang, you can write in any language, including Python.
+* Simple: a plugin just responds to RPC HTTP requests.
+* You can iterate quickly by changing the plugin at runtime.
+* You can get your plugin running today, no need to wait 3-5 months for an Argo software release.
+
+[Executor plugins](executor_plugins.md) written and installed by both users and admins.
+
+## Configuration
+
+Plugins are disabled by default. To enable them, start the controller with `ARGO_PLUGINS=true`, e.g.
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: workflow-controller
+spec:
+  template:
+    spec:
+      containers:
+        - name: workflow-controller
+          env:
+            - name: ARGO_PLUGINS
+              value: "true"
+```
+
+## Considerations
+
+### Security
+
+No security context is enforced, but is strong encouraged you configure your container not to run as root.
+
+Plugins can run at an non-localhost address, but this is very strongly discouraged. Within a pod, network traffic cannot
+be observed.
+
+### Resources
+
+No resource requests or limits are enforced (as they move override system defaults), but you are strongly encouraged to
+set these.
+
+### Failure Modes
+
+A plugin may fail as follows:
+
+* Connection/socket problems.
+* Timeout (1s for controller plugins, 30s for executor plugins).
+* Transient error.
+* 4xx or 5xx error:
+    * 404 error - endpoint will not be invoked again.
+    * 503 error - considered a transient error.
+* Multiple invocations of the same plugin take too long.
+
+Transient errors are retried, all other errors are considered fatal.
+
+Fatal errors are typically contained as follows:
+
+* For node lifecycle hooks, the node will error. The workflow therefore may fail.
```

Comment:
Does this PR include node lifecycle hooks? I think this is the first mention of them in this doc.